src/testnet/burnchain.rs

```diff
@@ -0,0 +1,169 @@
+use std::sync::mpsc;
+use std::thread;
+use std::time;
+use std::sync::{Arc, Mutex};
+
+use super::{Config};
+
+use burnchains::{Burnchain, BurnchainBlockHeader, BurnchainHeaderHash, BurnchainBlock, Txid};
+use burnchains::bitcoin::BitcoinBlock;
+use chainstate::burn::db::burndb::{BurnDB};
+use chainstate::burn::{BlockSnapshot};
+use chainstate::burn::operations::{BlockstackOperationType};
+use util::hash::Sha256Sum;
+
+/// BurnchainSimulator is simulating a simplistic burnchain.
+/// 
+/// When calling start(), a thread is being detached, looping, sleeping and firing some events, and a tuple (tx, rx) 
+/// is being returned so that the caller can be receiving new blocks, and transmitting ops in a simulated mem_pool.
+pub struct BurnchainSimulator {
+    mem_pool: Arc<Mutex<Vec<BlockstackOperationType>>>,
+    db: Option<Arc<Mutex<BurnDB>>>,
+}
+
+pub struct BurnchainState {
+    pub chain_tip: BlockSnapshot,
+    pub ops: Vec<BlockstackOperationType>,
+    pub db: Arc<Mutex<BurnDB>>
+}
+
+impl BurnchainSimulator {
+
+    pub fn new() -> Self {
+        Self {
+            mem_pool: Arc::new(Mutex::new(vec![])),
+            db: None,
+        }
+    }
+    
+    pub fn start(&mut self, config: &Config) -> (mpsc::Receiver<BurnchainState>, mpsc::Sender<BlockstackOperationType>) {
+        let (block_tx, block_rx) = mpsc::channel();
+        
+        let path = config.burnchain_path.clone();
+        let chain = config.chain.clone();
+        let name = config.testnet_name.clone();
+        let block_time = time::Duration::from_millis(config.burnchain_block_time);
+
+        let ops_dequeuing = Arc::clone(&self.mem_pool);
+
+        let db = match BurnDB::connect(&path, 0, &BurnchainHeaderHash([0u8; 32]), true) {
+            Ok(db) => Arc::new(Mutex::new(db)),
+            Err(_) => panic!("Error while connecting to burnchain db")
+        };
+
+        let burn_db = Arc::clone(&db);
+        self.db = Some(db);
+
+        thread::spawn(move || {
+
+            let chain = match Burnchain::new(&path, &chain, &name) {
+                Ok(res) => res,
+                Err(_) => panic!("Error while instantiating burnchain")
+            };
+
+            let mut chain_tip = {
+                let db = burn_db.lock().unwrap();
```

Comment:
Is the `Arc<BurnDB>` shared with any other threads?  I.e. could this `lock()` panic?

Reply:
2 threads are sharing the `Mutex<BurnDB>`:
- the burnchain, producing burnchain blocks
- the runloop, consuming and burnchain blocks / state.

If there's contention when trying to acquire the lock, then the latest thread attempting to acquire the `lock` will be blocking, and wait for the first lock to go out of scope. The `Mutex<BurnDB>` is the only shared resource shared between the 2 threads, so we should not end up with dead-locks.
When unwrapping the lock, the error that can be returned is (quoting the doc https://doc.rust-lang.org/std/sync/struct.Mutex.html#errors):
> If another user of this mutex panicked while holding the mutex, then this call will return an error once the mutex is acquired.

Which - I think - is the behavior we would expect.

Reply:
Gotcha.  This is fine then :+1: